sql: refactor internal executor factory to accept init function

adityamaru opened the pull request:
This change refactors the InternalExecutorFactory to take in
a closure that allows injecting external state into a new internal
executor.

Release note: None

adityamaru committed:
sql: refator internal executor factory to accept init function

This change refactors the InternalExecutorFactory to take in
a closure that allows injecting external state into a new internal
executor.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73193)
<!-- Reviewable:end -->

adityamaru committed:
sql,*ccl: remove InternalExecutor from ExecCfg

In a bid to make the InternalExecutor more tightly bound
with certain conn executor state such as transaction, session
data, extraTxnState (including descs collection), this patch
removes the `InternalExecutor` field from the executor config.
In this way consumers of the IE cannot just run queries off of
this free floating instance of the IE, but instead have to
initialize a new IE in the context of the surrounding txn.

The `InternalExecFactory` currently takes a closure that allows
for injecting necessary state into a newly created IE, but one
could imagine a future where we use typed parameters to force
all users to bind the IE with the surrounding state it is being
used in.

Release note: None
